### ISSUE TITLE
Investigate feasibility of zero-allocation protobuf marshalers

### ISSUE THREAD

**Issue by jack-berg** [open-telemetry/opentelemetry-java#5727]
The [OTLP protobuf marshalers](https://github.com/open-telemetry/opentelemetry-java/tree/main/exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/otlp) maintain a bunch of state as they transform `SpanData`, `MetricData`, `LogRecordData` to an intermediate representation before serializing. We should investigate whether we can eliminate this state to produce zero allocation marshalers.

Discussed in #5105 and in [discussion #5724](https://github.com/open-telemetry/opentelemetry-java/discussions/5724).

**Comment by asafm** [comment]
Protocol Buffers forces you to know the size of an embedded message before you write its binary encoding. 
This means you must:
1. Encode the embedded message to a temporary buffer
2. Grab the length of the buffer
3. Write the size of the buffer
4. Write the buffer (encoding of the message).

The intermediate classes' primary goal (i.e., `Marshaler*`) is to contain the size for each. Another goal is mapping the protobuf schema since the schema and the `MetricData` are not entirely 1:1 mapping.

I spent two days and had yet to do the serialization in a streaming fashion; hence, the only way out I saw was the reusable objects and object pooling. A very rough idea is at https://github.com/open-telemetry/opentelemetry-java/pull/6171.
